source/MaterialXTest/MaterialXCore/Node.cpp

```diff
@@ -609,97 +609,123 @@ TEST_CASE("Node Definition Creation", "[nodedef]")
     REQUIRE(graph);
     if (graph)
     {
+        // Add some input interfaces to the graph
+        for (auto node : graph->getNodes())
+        {
+            for (mx::InputPtr input : node->getInputs())
+            {
+                if (!input->getConnectedNode())
+                {
+                    const std::string relativePath = node->getName() + "/" + input->getName();
+                    const std::string interfaceName = graph->createValidChildName(relativePath);
+                    mx::InputPtr interfaceInput = graph->addInterfaceName(relativePath, interfaceName);
+                    REQUIRE(interfaceInput);
+                    if (interfaceInput)
+                    {
+                        interfaceInput->setAttribute(mx::PortElement::UI_NAME_ATTRIBUTE, node->getName() + " " + input->getName());
+                        interfaceInput->setAttribute(mx::PortElement::UI_FOLDER_ATTRIBUTE, "Common");
+                    }
+                }
+            }
+        }
+
         const std::string VERSION1 = "1.0";
         const std::string GROUP = "adjustment";
         bool isDefaultVersion = false;
         const std::string NODENAME = graph->getName();
 
-        // Duplicate the graph and then make the duplicate a nodedef nodegraph
+        // 1. Create a new functional graph and definitiion from a compound graph
         std::string newNodeDefName = doc->createValidChildName("ND_" + graph->getName());
         std::string newGraphName = doc->createValidChildName("NG_" + graph->getName());
-        mx::NodeDefPtr nodeDef = doc->addNodeDefFromGraph(graph, newNodeDefName, NODENAME, VERSION1, isDefaultVersion, GROUP, newGraphName);
+
+        mx::DefinitionOptions options;
+        options.categoryString = NODENAME;
+        options.compoundGraph = graph;
+        options.docString = mx::EMPTY_STRING;
+        options.isDefaultVersion = isDefaultVersion;
+        options.namespaceString = mx::EMPTY_STRING;
+        options.nodeGroupString = GROUP;
+        options.useNamespace = false;
+        options.useVersion = true;
+        options.versionString = VERSION1;
+
+        options.newNodeDefName = newNodeDefName;
+        options.newNodeGraphName = newGraphName;
+
+        mx::NodeDefPtr nodeDef = doc->createDefinition(options);
```

Comment:
This seems like a great use case for the new `createDefinition` functionality, so let's focus on this one for now.  In this situation, I would imagine it's clearer and more straightforward for a client to call the following:

```
mx::NodeDefPtr nodeDef = doc->createDefinition(graph);
nodeDef->setVersion("1.0");
nodeDef->setIsDefaultVersion(true);
```

This makes it clear that nearly all of the required data (e.g. name, category) is being derived from the NodeGraph itself, with only the versioning data for the NodeDef being stated explicitly by the client.

What are the benefits that you see in the DefinitionOptions class for this use case?  I could see it being useful if a single DefinitionOptions were constructed once and then reused multiple times, but that doesn't seem like a common situation that clients would encounter in practice.

Very interested in your thoughts on this!

Reply:
> This seems like a great use case for the new `createDefinition` functionality, so let's focus on this one for now. In this situation, I would imagine it's clearer and more straightforward for a client to call the following:
> 
> ```
> mx::NodeDefPtr nodeDef = doc->createDefinition(graph);
> nodeDef->setVersion("1.0");
> nodeDef->setIsDefaultVersion(true);
> ```
> 
> This makes it clear that nearly all of the required data (e.g. name, category) is being derived from the NodeGraph itself, with only the versioning data for the NodeDef being stated explicitly by the client.
> 
> What are the benefits that you see in the DefinitionOptions class for this use case? I could see it being useful if a single DefinitionOptions were constructed once and then reused multiple times, but that doesn't seem like a common situation that clients would encounter in practice.
> 
> Very interested in your thoughts on this!

1. For the first example, the main worry here is what if someone does this and then changes the version. This also involves the unique signature semantics being used. i.e. the proposal is the the identifier includes version. so you'd end up with something like `ND/NG_foo_<version>_....etc`
IMO, the user should not need to know where the information resides (whether it's on the graph or definition), though I agree that the options should probably be nodedef specific and not include the graph so it would be more of:
```
createDefinition(graph, nodedef_options)
```
I wanted to change this anyways as there is possible reuse when dealing with source code implementations -- i.e. the nodedef options are still valid. 

2. An example for the second case is if you want to create variants on definition. e.g. if you had nodegraphs with float, color3, vector3 etc outputs. Then you'd want consistent version, nodegroup, namespace, etc but the signature would
differ slightly depending on the nodegraph. e.g. You could end up with:
```
myspace:ND_foo_v1_float
myspace:ND_foo_v1_vector3
myspace:ND_ foo_v1_color3
etc...
```

Reply:
@kwokcb I'm not quite sure what you mean by "someone does this and changes the version".

Since the desired version is being stated just _once_ by the client, with no coupling between the NodeGraph and NodeDef, I don't see the benefit in stating this version inside of a `DefinitionOptions` class instead of stating it explicitly through a `setVersion` call.

What is the situation you're imagining, where using the `DefinitionOptions` structure would be more robust to version changes?

Reply:
I see where you're coming from, but I'm not quite as convinced that the `DefinitionOptions` structure adds clarity or robustness from the user's perspective.  To my mind it's just _hiding_ the details of what happens inside of `createDefinition`, and it seems clearer and more robust for the client to see these explicit steps in their code.  My hope is that `createDefinition` would handle all of the details that are implied by the input NodeGraph itself (e.g. name, category), but not try to handle details that the client needs to explicitly define (e.g. versioning, documentation, and so forth).

Reply:
Also my thinking is the emphasis of the API is on nodeef creation so you want the arguments available with a nodegraph is one source which affects the nodedef's characteristics, not so much that you create "some" nodedef with a nodegraph and then specify it's characteristics afterwards.

I think the flip-side question is how to create consistent definitions if the code logic is explicitly performed in separate steps. For example for "global" consistency the utility below exists. Having all args allows a way to take all the information to produce unique ids which are also used for Element existence checking. It would be difficult to promote this if examples show attributes being added after the definition creation. 
```
string generateIdentifier() const
    {
        if (!compoundGraph || categoryString.empty())
        {
            return EMPTY_STRING;
        }

        std::string identifier = categoryString;

        if (useVersion && !versionString.empty())
        {
            identifier += "_" + versionString;
        }

        for (OutputPtr output : compoundGraph->getOutputs())
        {
            const std::string& outputType = output->getType();
            identifier += "_" + outputType;
        }

        return identifier;
    }
```
For:

`namespace`: This needs to be set on the nodedef and nodegraph so the simple reasoning would be to just avoid mistakes. 

`nodegroup`: can be set after, but can change the definitions codegen behaviour, and as far as I've seen nobody thinks about this so will likely be forgotten. So folks can end up with unexpected shader gen results. For example
the houdini nodes are in the `houdini` group but if they were to add `pbr` nodes then the behaviour could be unknown.

category, `compound nodegraph`, & unique signatures for the nodedef and functional nodegraph should be consistent -- and this relies on the identifier generation.

`document` agreed can be done anytime, but again folks don't seem to think about adding it in, so most nodes are undocumented.

I just don't see anyone ever remembering this or they do how can you promote / enforce producing consistent definitions. 

BTW the existing `addNodeDefFromGraph()` already has most of the "desired" requirements. Perhaps we just keep this
with the logic fixes and add in the unique id generator as a util or embedded inside the API for id generation consistency :).  The only outlier is `namespace` which could be added for the aforementioned reason that I don't think
anyone will remember to set it on the graph and nodedef.  Which I think led to the struct creation...

Reply:
Omitting these changes to `addNodeDefFromGraph` sounds like a good strategy to me, and from a high level I would recommend that you make this pull request as minimal as it can possibly be.  That should definitely help me to provide a useful review with my limited GitHub development time, and hopefully it will make this change easier to maintain in the future as well.  :)

Reply:
Hi @jstone-lucasfilm 
I've created 2 smaller PRs. The original one which was just to update the API logic (no signature change) is #1403.
The UI integration into the graph editor is here #1404.

I will close this one out once I copy over all the PR docs.